src/components/Table/hooks.tsx

```diff
@@ -0,0 +1,258 @@
+import { useMemo, useReducer, useEffect, ReactNode } from "react";
+import {
+  ColumnByNamesType,
+  ColumnType,
+  TableState,
+  TableAction,
+  DataType,
+  UseTableReturnType,
+  UseTableOptionsType,
+  RowType,
+  HeaderType,
+  ColumnStateType,
+} from "./types";
+import { byTextAscending, byTextDescending } from "./utils";
+
+const sortByColumn = <T extends DataType>(
+  data: RowType<T>[],
+  sortColumn: string,
+  columns: ColumnStateType<T>[]
+): RowType<T>[] => {
+  let isAscending = null;
+  let sortedRows: RowType<T>[] = [...data];
+
+  columns.forEach((column) => {
+    // if the row was found
+    if (sortColumn === column.name) {
+      isAscending = column.sorted.asc;
+
+      if (column.sort) {
+        sortedRows = isAscending ? data.sort(column.sort) : data.sort(column.sort).reverse();
+        // default to sort by string
+      } else {
+        sortedRows = isAscending
+          ? data.sort(byTextAscending((object) => object.original[sortColumn]))
+          : data.sort(byTextDescending((object) => object.original[sortColumn]));
+      }
+    }
+  });
+
+  return sortedRows;
+};
+
+const getPaginatedData = <T extends DataType>(rows: RowType<T>[], perPage: number, page: number) => {
+  const start = (page - 1) * perPage;
+  const end = start + perPage;
+  return rows.slice(start, end);
+};
+
+const getColumnsByName = <T extends DataType>(columns: ColumnType<T>[]): ColumnByNamesType<T> => {
+  const columnsByName: ColumnByNamesType<T> = {};
+  columns.forEach((column) => {
+    const col: ColumnType<T> = {
+      id: column.id,
+      name: column.name,
+      label: column.label,
+    };
+
+    if (column.render) {
+      col.render = column.render;
+    }
+    col.hidden = column.hidden;
+    columnsByName[column.name] = col;
+  });
+
+  return columnsByName;
+};
+
+const sortDataInOrder = <T extends DataType>(data: T[], columns: ColumnType<T>[]): T[] => {
+  return data.map((row: T) => {
+    const newRow: DataType = {};
+    columns.forEach((column) => {
+      if (!(column.name in row)) {
+        throw new Error(`Invalid row data, ${column.name} not found`);
+      }
+      newRow[column.name] = row[column.name];
+    });
+    return newRow as T;
+  });
+};
+
+const makeRender = <T extends DataType, K>(
+  valueT: K,
+  render: (({ value, row }: { value: K; row: T }) => ReactNode) | undefined,
+  row: T
+) => {
+  return render ? () => render({ row, value: valueT }) : () => valueT;
+};
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const makeHeaderRender = (label: string, render: any) => {
+  return render ? () => render({ label }) : () => label;
+};
```

Comment:
A couple questions:

1. What is the purpose of this?

2. `makeRender` is for the body? Seems like a table header and table cell should have the same signature. Also if render is `any` this could throw an error if you pass a truthy value that is not a function.

Reply:
1. It's for customized cells. It allows us to define render method to customize cells
2. Yeah, `makeRender` is for body and this method is for internal usage so we shouldn't worry about not passing function